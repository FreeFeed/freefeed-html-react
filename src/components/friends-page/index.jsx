/* global CONFIG */
import { useEffect, useMemo, useCallback } from 'react';
import { Link } from 'react-router';
import { useSelector, useDispatch } from 'react-redux';
import cn from 'classnames';

import { Helmet } from 'react-helmet';
import ErrorBoundary from '../error-boundary';
import {
  combineAsyncStates,
  loadingAsyncState,
  successAsyncState,
  errorAsyncState,
  initialAsyncState,
} from '../../redux/async-helpers';
import { getAllSubscriptions, subscribers, blockedByMe } from '../../redux/action-creators';
import { Throbber } from '../throbber';
import { SignInLink } from '../sign-in-link';
import styles from './styles.module.scss';
import { Subscribers } from './subscribers';
import { Blocked } from './blocked';
import { Subscriptions } from './subscriptions';
import { Requests } from './requests';

const tabIds = ['', 'blocked', 'subscribers', 'requests'];

export const Friends = withLayout(function Friends({ router }) {
  const dispatch = useDispatch();
  const authenticated = useSelector((state) => state.authenticated);
  const thisUsername = useSelector((state) => state.user.username || null);
  const currentList = useSelector(
    (state) => state.routing.locationBeforeTransitions?.query.show || '',
  );
  const homeFeeds = useSelector((state) => state.homeFeeds);

  const allSubscriptionsStatus = useSelector((state) => state.allSubscriptionsStatus);
  const blockedByMeStatus = useSelector((state) => state.blockedByMeStatus);
  const subscribersStatus = useSelector((state) => toAsyncStatus(state.usernameSubscribers));

  const subscribersCount = useSelector((state) => state.usernameSubscribers.payload.length);
  const subscriptionsCount = useSelector((state) => state.allSubscriptions.length);
  const blockedCount = useSelector((state) => state.user.banIds?.length || 0);

  const incomingRequestsCount = useSelector(
    (state) => state.groupRequestsCount + state.userRequestsCount,
  );

  const dataStatus = useMemo(
    () => combineAsyncStates(allSubscriptionsStatus, blockedByMeStatus, subscribersStatus),
    [allSubscriptionsStatus, blockedByMeStatus, subscribersStatus],
  );

  const activeTab = useMemo(() => (tabIds.includes(currentList) ? currentList : ''), [currentList]);

  const onListChange = useCallback(
    (newList) => {
      newList !== currentList &&
        router.push({ ...router.location, query: newList ? { show: newList } : {} });
    },
    [currentList, router],
  );

  useEffect(() => {
    const validListValues = [...tabIds, ...homeFeeds.map((f) => f.id)];
    if (allSubscriptionsStatus.success && !validListValues.includes(currentList)) {
      router.replace({ ...router.location, query: {} });
    }
  }, [allSubscriptionsStatus.success, currentList, router, homeFeeds]);

  // Initial data loading
  useEffect(
    () =>
      void (
        authenticated &&
        (dispatch(getAllSubscriptions()),
        dispatch(subscribers(thisUsername)),
        dispatch(blockedByMe()))
      ),
    [authenticated, dispatch, thisUsername],
  );

  if (!authenticated) {
    return (
      <p>
        Please <SignInLink>sign in</SignInLink> to view this page.
      </p>
    );
  } else if (dataStatus.error) {
    return (
      <p className="alert alert-danger" role="alert">
        Cannot load page data: {dataStatus.errorText}
      </p>
    );
  } else if (!dataStatus.success) {
    return (
      <p>
        Loading data... <Throbber />
      </p>
    );
  }

  return (
    <>
      <Helmet
        title=""
        titleTemplate={`%s - Friends - ${CONFIG.siteTitle}`}
        defaultTitle={`Friends - ${CONFIG.siteTitle}`}
        defer={false}
      />

      <ul className={cn('nav nav-tabs', styles.topNav)} role="tablist">
        <Tab id="" activeId={activeTab}>
          Subscriptions <small className={styles.tabCounter}>{subscriptionsCount || ''}</small>
        </Tab>
        <Tab id="subscribers" activeId={activeTab}>
          Subscribers <small className={styles.tabCounter}>{subscribersCount || ''}</small>
        </Tab>
        <Tab id="blocked" activeId={activeTab}>
          Blocked users <small className={styles.tabCounter}>{blockedCount || ''}</small>
        </Tab>
        <Tab id="requests" activeId={activeTab}>
          Requests{' '}
          {incomingRequestsCount > 0 && (
            <>
              <small className={styles.tabCounter}>{incomingRequestsCount || ''}</small>
              <span className={styles.incomingRequestsMarker} />
            </>
          )}
        </Tab>
      </ul>

      <div role="tabpanel">
        {activeTab === 'subscribers' ? (
          <Subscribers />
        ) : activeTab === 'blocked' ? (
          <Blocked />
        ) : activeTab === 'requests' ? (
          <Requests />
        ) : (
          <Subscriptions listId={currentList} onListChange={onListChange} />
        )}
      </div>
    </>
  );
});

function withLayout(Component) {
  const wrapper = (props) => (
    <div className="box">
      <ErrorBoundary>
        <div className="box-header-timeline" role="heading">
          Friends
        </div>
        <div className="box-body">
          <Component {...props} />
        </div>
        <div className="box-footer" />
      </ErrorBoundary>
    </div>
  );
  wrapper.displayName = `Layout(${Component.displayName || Component.name || 'unnamed'})`;
  return wrapper;
}

function toAsyncStatus({ initial, isPending, errorString }) {
  if (initial) {
    return initialAsyncState;
  } else if (isPending) {
    return loadingAsyncState;
  } else if (errorString === '') {
    return successAsyncState;
  }
  return errorAsyncState(errorString);
}

function Tab({ id, activeId, children }) {
  const location = useSelector((state) => state.routing.locationBeforeTransitions);
  return (
    <li role="tab presentation" className={cn({ active: id === activeId })}>
      <Link to={{ ...location, query: id ? { show: id } : {} }}>{children}</Link>
    </li>
  );
}
