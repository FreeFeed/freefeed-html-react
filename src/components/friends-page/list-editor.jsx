import { memo, useMemo, useCallback, useState, useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import cn from 'classnames';
import { faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';

import {
  createHomeFeed,
  updateHomeFeed,
  deleteHomeFeed,
  updateHomeFeedReset,
  getAllSubscriptions,
} from '../../redux/action-creators';
import { doSequence } from '../../redux/async-helpers';
import { andJoin } from '../../utils/and-join';
import { UserPicture } from '../user-picture';
import { ButtonLink, useKeyboardEvents } from '../button-link';
import { Throbber } from '../throbber';
import { Icon } from '../fontawesome-icons';
import { OverlayPopup } from '../overlay-popup';
import styles from './list-editor.module.scss';

const SHOW_ALL_USERS = 'all';
const SHOW_ONLY_SELECTED_USERS = 'selected';
const SHOW_ONLY_NOT_SELECTED_USERS = 'notSelected';

export const ListEditor = memo(function ListEditor({
  listId = null,
  // close function receives an argument that is:
  // 1. ID of edited _or created_ list
  // 2. null or undefined if editor was closed with "Cancel" button
  close = () => null,
}) {
  const dispatch = useDispatch();
  const allSubscriptions = useSelector((state) => state.allSubscriptions);
  const allHomeFeeds = useSelector((state) => state.homeFeeds);
  const allUsers = useSelector((state) => state.users);
  const allSubscriptionsStatus = useSelector((state) => state.allSubscriptionsStatus);
  const submitStatus = useSelector((state) => state.crudHomeFeedStatus);

  // Fetch subscriptions and homefeeds if there are not fetched yet
  useEffect(
    () => void (allSubscriptionsStatus.initial && dispatch(getAllSubscriptions())),
    [allSubscriptionsStatus.initial, dispatch],
  );

  // Edited list (or null for the new one)
  const list = useMemo(
    () => allHomeFeeds.find((feed) => feed.id === listId),
    [allHomeFeeds, listId],
  );

  // Map of user's selection states: { [userId]: boolean }
  const [selected, setSelected] = useState({});
  // Fill the selection map when allSubscriptions data is ready
  useEffect(() => {
    const listId = list?.id;
    if (allSubscriptionsStatus.success && listId) {
      const sel = {};
      for (const { id, homeFeeds } of allSubscriptions) {
        if (homeFeeds.includes(listId)) {
          sel[id] = true;
        }
      }
      setSelected(sel);
    }
  }, [allSubscriptionsStatus.success, allSubscriptions, list?.id]);

  const totalCount = allSubscriptions.length;
  const selectedCount = useMemo(() => Object.values(selected).filter(Boolean).length, [selected]);
  const notSelectedCount = totalCount - selectedCount;

  const onSelect = useCallback((id, sel) => setSelected({ ...selected, [id]: sel }), [selected]);

  // Filter users by selection
  const [showMode, setShowMode] = useState(SHOW_ALL_USERS);
  const showAll = useCallback(() => setShowMode(SHOW_ALL_USERS), []);
  const showSelected = useCallback(() => setShowMode(SHOW_ONLY_SELECTED_USERS), []);
  const showNotSelected = useCallback(() => setShowMode(SHOW_ONLY_NOT_SELECTED_USERS), []);

  // Filter users by name
  const [nameFilter, setNameFilter] = useState('');
  const onFilterChange = useCallback(({ target }) => setNameFilter(target.value), []);
  const clearFilter = useCallback(() => setNameFilter(''), []);

  // List title edit
  const [listTitle, setListTitle] = useState('');
  const onListTitleChange = useCallback(({ target }) => setListTitle(target.value), []);

  // Set list title when data is loaded
  useEffect(() => {
    if (allSubscriptionsStatus.success) {
      setListTitle(list?.title || '');
    }
  }, [allSubscriptionsStatus.success, list?.title]);

  // Prepare users data
  const preparedUsers = useMemo(() => {
    return allSubscriptions.map(({ id, homeFeeds }) => ({
      ...allUsers[id],
      selected: !!selected[id],
      homeless: homeFeeds.length === 0,
    }));
  }, [allSubscriptions, allUsers, selected]);

  const usersToShow = useMemo(() => {
    const filter = nameFilter.trim().toLocaleLowerCase();
    return preparedUsers
      .filter((u) => {
        if (showMode === SHOW_ALL_USERS) {
          return true;
        }
        if (showMode === SHOW_ONLY_SELECTED_USERS && u.selected) {
          return true;
        }
        if (showMode === SHOW_ONLY_NOT_SELECTED_USERS && !u.selected) {
          return true;
        }
        return false;
      })
      .filter(
        (u) =>
          !filter ||
          u.username.toLocaleLowerCase().includes(filter) ||
          u.screenName.toLocaleLowerCase().includes(filter),
      )
      .sort((a, z) => a.username.localeCompare(z.username));
  }, [nameFilter, preparedUsers, showMode]);

  const homeless = useMemo(
    () => preparedUsers.filter((u) => u.homeless && !u.selected),
    [preparedUsers],
  );
  const returnListId = useRef(list?.id);

  // Handle submit
  const canSubmit = useMemo(
    () => allSubscriptionsStatus.success && listTitle.trim() !== '',
    [allSubscriptionsStatus.success, listTitle],
  );

  const doSubmit = useCallback(() => {
    if (!canSubmit) {
      return;
    }
    const listId = list && list.id;
    const data = {
      subscribedTo: Object.keys(selected).filter((id) => !!selected[id]),
    };

    if (!list || !list.isInherent) {
      data.title = listTitle;
    }
    doSequence(dispatch)(
      (dispatch) => dispatch(listId ? updateHomeFeed(listId, data) : createHomeFeed(data)),
      (_, { payload }) => (returnListId.current = payload.timeline.id),
    );
  }, [canSubmit, selected, dispatch, list, listTitle]);

  // Handle cancel
  const doCancel = useCallback(() => close(), [close]);

  // Handle delete
  const doDelete = useCallback(() => {
    const selectedIds = Object.keys(selected).filter((id) => !!selected[id]);
    if (
      !confirm(
        `Are you sure you want to delete this list?${
          selectedIds.length > 0
            ? `\nAll remaining list subscriptions will be moved to the main Home list.`
            : ``
        }`,
      )
    ) {
      return;
    }
    const listId = list?.id;

    doSequence(dispatch)(
      (dispatch) => dispatch(updateHomeFeed(listId, { subscribedTo: selectedIds })),
      (dispatch) => dispatch(deleteHomeFeed(listId)),
    );
  }, [dispatch, list?.id, selected]);

  // Close this dialog on success and pass returnListId to the close function
  useEffect(
    () =>
      void (submitStatus.success && (close(returnListId.current), dispatch(updateHomeFeedReset()))),
    [submitStatus.success, close, dispatch],
  );

  return (
    <OverlayPopup close={doCancel}>
      <section className={styles.wrapper} data-testid="list-editor">
        <header className={styles.header}>
          {list && list.isInherent ? (
            <h3>{list.title}</h3>
          ) : (
            <dl className={styles.topForm}>
              <dt>List title</dt>
              <dd>
                <input
                  type="text"
                  autoFocus={!list}
                  className="form-control input-lg"
                  placeholder={`Name your ${list ? '' : 'new '}list (required)`}
                  value={listTitle}
                  onChange={onListTitleChange}
                />
              </dd>
            </dl>
          )}
          <p>
            Show:{' '}
            <ButtonLinkX onClick={showAll} selected={showMode === SHOW_ALL_USERS}>
              All ({totalCount})
            </ButtonLinkX>
            {' - '}
            <ButtonLinkX onClick={showSelected} selected={showMode === SHOW_ONLY_SELECTED_USERS}>
              In the list ({selectedCount})
            </ButtonLinkX>
            {' - '}
            <ButtonLinkX
              onClick={showNotSelected}
              selected={showMode === SHOW_ONLY_NOT_SELECTED_USERS}
            >
              Not in the list ({notSelectedCount})
            </ButtonLinkX>
            {' - '}
            Filter by name:{' '}
            <input
              type="search"
              name="nameFilter"
              autoFocus={!!list}
              value={nameFilter}
              onChange={onFilterChange}
              className={cn('form-control input-sm', styles.filterInput)}
            />{' '}
            (<ButtonLink onClick={clearFilter}>clear</ButtonLink>)
          </p>
        </header>
        <main>
          <div className={styles.content}>
            {allSubscriptionsStatus.success ? (
              <UsersList users={usersToShow} onSelect={onSelect} highlight={nameFilter.trim()} />
            ) : allSubscriptionsStatus.error ? (
              <p className="alert alert-danger" role="alert">
                Error loading subscriptions: {allSubscriptionsStatus.errorText}
              </p>
            ) : (
              <p>Loading...</p>
            )}
          </div>
        </main>
        <footer className={styles.footer}>
          <button
            className={cn('btn btn-primary ', (!canSubmit || submitStatus.loading) && 'disabled')}
            type="submit"
            onClick={doSubmit}
          >
            {list
              ? submitStatus.loading
                ? `Saving changes...`
                : `Save changes`
              : submitStatus.loading
              ? 'Creating list...'
              : 'Create list'}
          </button>
          <button className="btn btn-link" type="reset" onClick={doCancel}>
            Cancel
          </button>
          {submitStatus.loading && <Throbber />}
          {list && !list.isInherent && (
            <button
              className={cn('btn btn-danger pull-right', submitStatus.loading && 'disabled')}
              type="button"
              onClick={doDelete}
            >
              Delete list
            </button>
          )}
          {submitStatus.error && (
            <p className="alert alert-danger" role="alert">
              {submitStatus.errorText}
            </p>
          )}
          {homeless.length > 0 && (
            <div className={cn(styles.submitWarning, 'text-muted')}>
              <Icon icon={faExclamationTriangle} />{' '}
              {homeless.length < 5
                ? andJoin(homeless.map((u) => `@${u.username}`))
                : `${homeless.slice(0, 3).map((u) => `@${u.username}`)} and ${
                    homeless.length - 3
                  } more users `}
              {homeless.length === 1 ? 'is' : 'are'} not in any of your friend lists.
            </div>
          )}
        </footer>
      </section>
    </OverlayPopup>
  );
});

const UsersList = memo(function UsersList({ users, onSelect, highlight }) {
  if (users.length === 0) {
    return <p>No users here</p>;
  }
  return (
    <ul className={styles.list}>
      {users.map((user) => (
        <UserCell key={user.id} user={user} onSelect={onSelect} highlight={highlight} />
      ))}
    </ul>
  );
});

const UserCell = memo(function UserCell({ user, onSelect, highlight }) {
  const kbdHandlers = useKeyboardEvents(
    useCallback(() => onSelect(user.id, !user.selected), [user, onSelect]),
  );

  return (
    <li
      className={cn(styles.cell, user.selected && styles.cellSelected)}
      role="checkbox"
      aria-checked={user.selected}
      tabIndex={0}
      {...kbdHandlers}
    >
      <UserPicture user={user} withLink={false} />
      <div>
        <div className={styles.screenName}>{hl(user.screenName, highlight)}</div>
        <div className={styles.username}>@{hl(user.username, highlight)}</div>
        {user.homeless && (
          <Icon
            icon={faExclamationTriangle}
            className={styles.homelessMark}
            title={`This ${user.type} is in no lists`}
          />
        )}
      </div>
    </li>
  );
});

function hl(text, search) {
  const lowercaseText = text.toLocaleLowerCase();
  const lowercaseSearch = search.toLocaleLowerCase();
  if (lowercaseSearch === '' || !lowercaseText.includes(lowercaseSearch)) {
    return text;
  }
  const result = [];

  const highlightFrom = lowercaseText.indexOf(lowercaseSearch);
  const highlightTo = highlightFrom + search.length;
  result.push(text.slice(0, highlightFrom));
  result.push(
    <mark className={styles.mark} key={`${text}__${result.length}`}>
      {text.slice(highlightFrom, highlightTo)}
    </mark>,
  );
  result.push(text.slice(highlightTo));

  return result;
}

function ButtonLinkX({ selected = false, onClick, ...props }) {
  if (selected) {
    props.className = cn(styles.currentLink, props.className);
    return <span className={styles.currentLink} {...props} />;
  }
  return <ButtonLink onClick={onClick} {...props} />;
}
