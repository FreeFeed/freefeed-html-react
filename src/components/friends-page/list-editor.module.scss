@import '../../../styles/shared/mixins';
@import '../../../styles/helvetica/dark-vars.scss';

.wrapper {
  display: flex;
  flex-direction: column;
  max-height: 100%;
  width: 900px;
  max-width: 100%;
  padding: 1em;

  @media (max-width: 600px) {
    margin-top: 2em;
  }
}

.header {
  flex: 0 0;
}

.header h3 {
  margin-top: rem(3px);
}

.footer {
  flex: 0 0;
  padding-top: 1em;
}

.content {
  flex: 1;
  height: 50vh;
  overflow: auto;
  border-top: 1px solid #ccc;
  border-bottom: 1px solid #ccc;

  :global(.dark-theme) & {
    border-top-color: $separator-color;
    border-bottom-color: $separator-color;
  }
}

.list {
  display: flex;
  flex-wrap: wrap;
  align-content: flex-start;
  list-style-type: none;
  padding: 0;
}

.cell {
  margin: rem(4px);
  border-radius: rem(4px);
  width: 200px;
  word-break: break-word;
  font-size: rem(12px);
  display: grid;
  grid-auto-flow: column;
  gap: rem(4px);
  grid-template-columns: 0fr 1fr;
  padding: rem(4px);
  font-weight: normal;
  cursor: pointer;
  position: relative;

  &:hover {
    background: #eee;

    :global(.dark-theme) & {
      background: $bg-color-lightest;
    }
  }
}

.cellSelected,
.cellSelected:hover {
  background: hsl(200, 70%, 90%);

  :global(.dark-theme) & {
    color: $text-color-lighter;
    background: $bg-highlight-color;
  }
}

.username {
  color: #999;

  :global(.dark-theme) & {
    color: $text-color-darker;
  }
}

.mark {
  padding: 0;
  color: inherit;
  background-color: #ff0;

  :global(.dark-theme) & {
    color: $text-color-darker;
    background-color: $accent-color;
  }
}

.currentLink {
  font-weight: bold;
}

.topForm {
  display: grid;
  grid-template-columns: 0fr 1fr;
  white-space: nowrap;
  gap: 0.5em;
  max-width: 500px;
  margin-bottom: 1em;

  dt {
    font: inherit;
    align-self: center;
  }
}

.homelessMark {
  position: absolute;
  bottom: 4px;
  right: 4px;
}

.submitWarning {
  margin-top: 0.5em;
}

.filterInput {
  display: inline-block;
  width: 180px;
}
