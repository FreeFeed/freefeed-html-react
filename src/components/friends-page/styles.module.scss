@import '../../../styles/shared/mixins';

.topNav {
  margin-top: 2em;
}

.listSelector {
  display: inline-block;
  width: auto;
}

.usersGrid {
  display: flex;
  flex-wrap: wrap;
  list-style-type: none;
  padding: 0;
}

.userCell {
  width: 110px;
  min-height: 90px;
  text-align: center;
  word-wrap: break-word;
  font-size: rem(12px);
  line-height: rem(16px);
  margin: 0 rem(6px) rem(20px) rem(6px);
}

.userCellPlaceholder {
  @extend .userCell;

  $border: rgba(0, 0, 0, 0.5);
  $bg: #f6fafd;

  // Draw square as a userpic placeholder
  background-image: linear-gradient(
      to left,
      $border 0,
      transparent 1px,
      transparent 49px,
      $border 50px
    ),
    linear-gradient($border 0, $bg 1px, $bg 49px, $border 50px);
  background-repeat: no-repeat;
  background-position: 50% 0%;
  background-size: 50px 50px;
}

.listWithoutGroups {
  margin-top: 2em;
}

.listSelectorContainer {
  margin-top: 1em;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
}

.incomingRequestsMarker {
  color: #fd3c3c;
  position: relative;
  display: inline-block;

  &::after {
    position: absolute;
    top: -1em;
    right: -0.5em;
    content: '';
    border-radius: 50%;
    width: rem(6px);
    height: rem(6px);
    background-color: currentColor;
  }
}

.tabCounter {
  &:empty {
    display: none;
  }

  &::before {
    content: '(';
  }
  &::after {
    content: ')';
  }
}

.usersList {
  list-style-type: none;
  padding: 0;
  max-width: 400px;
}

.usersListItem {
  padding: 0;
  display: grid;
  gap: 0.5em;
  grid-auto-flow: column;
  grid-template-columns: 1fr 0fr;
  margin-bottom: 1em;
}

.btnGroup {
  display: inline-grid;
  grid-auto-flow: column;
  gap: 0.33em;
}

.userHorCell {
  display: grid;
  grid-auto-flow: column;
  grid-template-columns: 0fr 1fr;
  gap: 0.5em;
  white-space: nowrap;
}

.requestsSection {
  margin-bottom: 2em;
}
