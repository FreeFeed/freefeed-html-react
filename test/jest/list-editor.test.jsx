/* global describe, it, expect, jest, beforeEach */
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import { createStore } from 'redux';
import * as reactRedux from 'react-redux';

import { ListEditor } from '../../src/components/friends-page/list-editor';

const LIST_ID = 'list-id';

const defaultState = {
  allSubscriptions: [
    { id: 'u1', homeFeeds: [LIST_ID] },
    { id: 'u2', homeFeeds: [LIST_ID, 'some-other-list'] },
    { id: 'u3', homeFeeds: ['some-other-list'] },
    { id: 'u4', homeFeeds: [] },
    { id: 'g1', homeFeeds: [LIST_ID] },
  ],
  homeFeeds: [
    {
      id: LIST_ID,
      title: 'My amazing list',
      isInherent: false,
    },
  ],
  users: {
    u1: { id: 'u1', type: 'user', username: 'in-list-1', screenName: 'In list' },
    u2: { id: 'u2', type: 'user', username: 'in-list-2', screenName: 'In multiple lists' },
    u3: { id: 'u3', type: 'user', username: 'not-in-list', screenName: 'In Some Other List' },
    u4: { id: 'u4', type: 'user', username: 'homeless', screenName: 'Homeless' },
    u5: { id: 'u5', type: 'user', username: 'unrelated', screenName: 'Not subscribed' },
    g1: { id: 'g1', type: 'group', username: 'group', screenName: 'A group' },
  },
  allSubscriptionsStatus: { success: true },
  crudHomeFeedStatus: { success: true },
};

const renderListEditor = (props = {}, options = {}) => {
  const { Provider } = reactRedux;
  const dummyReducer = (state) => state;
  const store = createStore(dummyReducer, defaultState);

  const defaultProps = {
    listId: LIST_ID,
    closeEditor: () => {},
  };

  const rendered = render(
    <Provider store={store}>
      <ListEditor {...defaultProps} {...props} />
    </Provider>,
    options,
  );

  return {
    ...rendered,
    rerender: (props = {}, options = {}) =>
      renderListEditor(props, { container: rendered.container, ...options }),
  };
};

describe('ListEditor', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
    useSelectorMock.mockImplementation((selector) => selector(defaultState));
    useDispatchMock.mockImplementation(() => () => {});
  });

  it("Renders a list and doesn't blow up", () => {
    renderListEditor();
    const listEditor = screen.getByTestId('list-editor');
    expect(listEditor).toMatchSnapshot();
  });

  it('Add a user to the list by clicking on it', () => {
    renderListEditor();
    userEvent.click(screen.getByText(/Not in the list/));
    expect(screen.getByText('Homeless')).toBeDefined();
    expect(screen.getByText('In Some Other List')).toBeDefined();
    expect(screen.getByText(/not in any of your friend lists/)).toBeDefined();
    expect(screen.queryByText('In list')).toBeNull();
    userEvent.click(screen.getByText('Homeless'));
    expect(screen.queryByText('Homeless')).toBeNull();
    expect(screen.queryByText(/not in any of your friend lists/)).toBeNull();
  });

  it('Searches for users', () => {
    renderListEditor();
    userEvent.type(screen.getByRole('searchbox'), 'LIST');
    expect(screen.getByRole('list')).toMatchSnapshot();
  });
});
