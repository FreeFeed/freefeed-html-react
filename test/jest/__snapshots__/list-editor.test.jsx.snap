// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListEditor Renders a list and doesn't blow up 1`] = `
<section
  class="wrapper"
  data-testid="list-editor"
>
  <header
    class="header"
  >
    <dl
      class="topForm"
    >
      <dt>
        List title
      </dt>
      <dd>
        <input
          class="form-control input-lg"
          placeholder="Name your list (required)"
          type="text"
          value="My amazing list"
        />
      </dd>
    </dl>
    <p>
      Show:
       
      <span
        class="currentLink"
      >
        All (
        5
        )
      </span>
       - 
      <a
        aria-disabled="false"
        role="button"
        tabindex="0"
      >
        In the list (
        3
        )
      </a>
       - 
      <a
        aria-disabled="false"
        role="button"
        tabindex="0"
      >
        Not in the list (
        2
        )
      </a>
       - 
      Filter by name:
       
      <input
        class="form-control input-sm filterInput"
        name="nameFilter"
        type="search"
        value=""
      />
       
      (
      <a
        aria-disabled="false"
        role="button"
        tabindex="0"
      >
        clear
      </a>
      )
    </p>
  </header>
  <main>
    <div
      class="content"
    >
      <ul
        class="list"
      >
        <li
          aria-checked="true"
          class="cell cellSelected"
          role="checkbox"
          tabindex="0"
        >
          <div
            class="picture"
          >
            <img
              alt="Profile picture of group"
              height="50"
              width="50"
            />
          </div>
          <div>
            <div
              class="screenName"
            >
              A group
            </div>
            <div
              class="username"
            >
              @
              group
            </div>
          </div>
        </li>
        <li
          aria-checked="false"
          class="cell"
          role="checkbox"
          tabindex="0"
        >
          <div
            class="picture"
          >
            <img
              alt="Profile picture of homeless"
              height="50"
              width="50"
            />
          </div>
          <div>
            <div
              class="screenName"
            >
              Homeless
            </div>
            <div
              class="username"
            >
              @
              homeless
            </div>
            <svg
              aria-hidden="true"
              class="homelessMark fa-icon fa-icon-fas-exclamation-triangle"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                This user is in no lists
              </title>
              <path
                d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                fill="currentColor"
              />
            </svg>
          </div>
        </li>
        <li
          aria-checked="true"
          class="cell cellSelected"
          role="checkbox"
          tabindex="0"
        >
          <div
            class="picture"
          >
            <img
              alt="Profile picture of in-list-1"
              height="50"
              width="50"
            />
          </div>
          <div>
            <div
              class="screenName"
            >
              In list
            </div>
            <div
              class="username"
            >
              @
              in-list-1
            </div>
          </div>
        </li>
        <li
          aria-checked="true"
          class="cell cellSelected"
          role="checkbox"
          tabindex="0"
        >
          <div
            class="picture"
          >
            <img
              alt="Profile picture of in-list-2"
              height="50"
              width="50"
            />
          </div>
          <div>
            <div
              class="screenName"
            >
              In multiple lists
            </div>
            <div
              class="username"
            >
              @
              in-list-2
            </div>
          </div>
        </li>
        <li
          aria-checked="false"
          class="cell"
          role="checkbox"
          tabindex="0"
        >
          <div
            class="picture"
          >
            <img
              alt="Profile picture of not-in-list"
              height="50"
              width="50"
            />
          </div>
          <div>
            <div
              class="screenName"
            >
              In Some Other List
            </div>
            <div
              class="username"
            >
              @
              not-in-list
            </div>
          </div>
        </li>
      </ul>
    </div>
  </main>
  <footer
    class="footer"
  >
    <button
      class="btn btn-primary "
      type="submit"
    >
      Save changes
    </button>
    <button
      class="btn btn-link"
      type="reset"
    >
      Cancel
    </button>
    <button
      class="btn btn-danger pull-right"
      type="button"
    >
      Delete list
    </button>
    <div
      class="submitWarning text-muted"
    >
      <svg
        aria-hidden="true"
        class="fa-icon fa-icon-fas-exclamation-triangle"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          fill="currentColor"
        />
      </svg>
       
      @homeless
      is
       not in any of your friend lists.
    </div>
  </footer>
</section>
`;

exports[`ListEditor Searches for users 1`] = `
<ul
  class="list"
>
  <li
    aria-checked="true"
    class="cell cellSelected"
    role="checkbox"
    tabindex="0"
  >
    <div
      class="picture"
    >
      <img
        alt="Profile picture of in-list-1"
        height="50"
        width="50"
      />
    </div>
    <div>
      <div
        class="screenName"
      >
        In 
        <mark
          class="mark"
        >
          list
        </mark>
        
      </div>
      <div
        class="username"
      >
        @
        in-
        <mark
          class="mark"
        >
          list
        </mark>
        -1
      </div>
    </div>
  </li>
  <li
    aria-checked="true"
    class="cell cellSelected"
    role="checkbox"
    tabindex="0"
  >
    <div
      class="picture"
    >
      <img
        alt="Profile picture of in-list-2"
        height="50"
        width="50"
      />
    </div>
    <div>
      <div
        class="screenName"
      >
        In multiple 
        <mark
          class="mark"
        >
          list
        </mark>
        s
      </div>
      <div
        class="username"
      >
        @
        in-
        <mark
          class="mark"
        >
          list
        </mark>
        -2
      </div>
    </div>
  </li>
  <li
    aria-checked="false"
    class="cell"
    role="checkbox"
    tabindex="0"
  >
    <div
      class="picture"
    >
      <img
        alt="Profile picture of not-in-list"
        height="50"
        width="50"
      />
    </div>
    <div>
      <div
        class="screenName"
      >
        In Some Other 
        <mark
          class="mark"
        >
          List
        </mark>
        
      </div>
      <div
        class="username"
      >
        @
        not-in-
        <mark
          class="mark"
        >
          list
        </mark>
        
      </div>
    </div>
  </li>
</ul>
`;
